Multibyte character in array key makes alignment incorect

kubawerlos opened the pull request:
Running:

```bash
php php-cs-fixer fix --rules=mb_str_functions src/Fixer/Operator/BinaryOperatorSpacesFixer.php
```

would solve a problem, but is that what we want?

keradus commented:
ping @SpacePossum

SpacePossum commented:
Ditching all the `utf8_decode` from `src/Fixer/Operator/BinaryOperatorSpacesFixer.php` resolves the issue, but I'm sure those were added for some reason. Now we have to find out why... :|

kubawerlos commented:
Got me thinking about more general solution - what about adding `mb_str_functions` to the `.php_cs.dist`? This would require only small change in `EncodingFixer` (`mb_substr` 2nd parameter `3` -> `1`)

@SpacePossum @keradus thoughts?

SpacePossum commented:
I wouldn't reply on `mb_` everywhere as it is not needed everywhere.
If it helps here than lets check it for this fixer only and worry about the rest of the code base later :)

kubawerlos commented:
@SpacePossum I made smallest working fix for this issue.

SpacePossum commented on src/Fixer/Operator/BinaryOperatorSpacesFixer.php:
```diff
@@ -760,7 +760,7 @@ private function replacePlaceholders(Tokens $tokens, $alignStrategy)
                         }
 
-                        $lines[$index] = $before.substr($lines[$index], $currentPosition);
+                        $lines[$index] = $before.mb_substr($lines[$index], $currentPosition);
```
We don't require `mb_string` extension nor a polyfill for it (we only suggest it in the composer.json). Therefore this might not work for everyone. 
Not sure if we want to start requiring it here now, can we work around it in a reasonable matter?

kubawerlos commented:
@SpacePossum @julienfalque updated again.

So as I understand: there were 3 calls of `utf8_decode` in the fixer. I have added test case "align correctly with multibyte characters in array key" basing on similar one existing in code and it was failing either with 3 calls.

When I removed the one in this PR it's all working. Removing other `utf8_decode` make tests failing agains so maybe the removed one wasn't necessary.

kubawerlos commented:
ping @SpacePossum

SpacePossum commented:
@kubawerlos looking good :)

please add/update the tests to:

```php
                '<?php $a = 1 + 2; $b = array(
                    13 => 3,
                    4  => 3,
                    5  => 2,
                );
                $a = 12 + 1;
                $a = 13 + 41;
                ',
                '<?php $a    =   1   +    2; $b = array(
                    13 =>3,
                    4  =>  3,
                    5=>2,
                );
                $a = 12   +  1;
                $a = 13+41;
                ',
                ['default' => BinaryOperatorSpacesFixer::ALIGN_SINGLE_SPACE_MINIMAL],
```

```php
            'align correctly with multibyte characters in array key' => [
                '<?php
                    $inflect_male = array(
                        "aitė\b" => "øasø",
                        "ytė\b"  => "øisø",
                        "iūtė\b" => "øiusø",
                        "utė\b"  => array(
                            "aitė\b" => "øas",
                            "ytė\b"  => "øis",
                            "iūtė\b" => $øøius,
                            "utė\b"  => "us",
                        ),
                    );',
                '<?php
                    $inflect_male = array(
                        "aitė\b" => "øasø",
                        "ytė\b" => "øisø",
                        "iūtė\b" => "øiusø",
                        "utė\b" => array(
                            "aitė\b" => "øas",
                            "ytė\b" => "øis",
                            "iūtė\b" => $øøius,
                            "utė\b"  =>     "us",
                        ),
                    );',
                ['operators' => ['=>' => BinaryOperatorSpacesFixer::ALIGN_SINGLE_SPACE]],
            ],
```

these should pass as well.
After that please rebase and than I'm :+1:  for merging this, thanks!

kubawerlos commented:
@SpacePossum done

SpacePossum commented:
@keradus can you check `circleci` failure, I've restarted it but it doesn't help :(

keradus commented:
@SpacePossum circleCi failure is not related to this PR, it happens to other PRs as well.
Please ignore it for this PR in reviewing (and, if have time, investigate it separately)

keradus commented:
Thank you @kubawerlos.
